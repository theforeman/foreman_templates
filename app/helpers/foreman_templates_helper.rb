module ForemanTemplatesHelper
  def edit_paths
    Template.subclasses.reduce({}) do |memo, subclass|
      memo.tap do |acc|
        path_part = "#{subclass.name.underscore.pluralize}_path"
        acc[subclass.name] = "#{Rails.application.routes.url_helpers.public_send(path_part)}/:id/edit"
      end
    end
  end
end
