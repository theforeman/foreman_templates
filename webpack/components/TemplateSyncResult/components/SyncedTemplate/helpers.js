import React from 'react';
import { isEmpty } from 'lodash';
import { Icon } from 'patternfly-react';

import IconInfoItem from './IconInfoItem';
import EmptyInfoItem from './EmptyInfoItem';
import StringInfoItem from './StringInfoItem';

export const itemIteratorId = (template, attr) => `${template.name}-${attr}`;

export const additionalInfo = template => {
  const infoAttrs = [
    'locked',
    'snippet',
    'humanizedClassName',
    'kind',
    'templateFile',
  ];

  return infoAttrs.map(attr => {
    const key = itemIteratorId(template, attr);

    const classNameMap = { Ptable: 'Partition Table' };

    if (!template[attr]) {
      return <EmptyInfoItem template={template} attr={attr} key={key} />;
    }

    switch (attr) {
      case 'locked':
        return (
          <IconInfoItem
            template={template}
            attr={attr}
            iconName="lock"
            tooltipText="Locked"
            key={key}
          />
        );
      case 'snippet':
        return (
          <IconInfoItem
            template={template}
            attr={attr}
            iconName="check"
            tooltipText="Snippet"
            key={key}
          />
        );
      case 'humanizedClassName':
        return (
          <StringInfoItem
            template={template}
            attr={attr}
            translate
            mapAttr={(templateObj, attribute) =>
              classNameMap[templateObj[attribute]]
                ? classNameMap[templateObj[attribute]]
                : templateObj[attribute]
            }
            key={key}
          />
        );
      case 'kind':
        return <StringInfoItem template={template} attr={attr} key={key} />;
      case 'templateFile':
        return (
          <StringInfoItem template={template} attr={attr} key={key} elipsed />
        );
      default:
        return '';
    }
  });
};

export const itemLeftContentIcon = template => {
  const iconName = isEmpty(aggregatedErrors(template))
    ? 'ok'
    : 'error-circle-o';
  return <Icon name={iconName} size="sm" type="pf" />;
};

export const templateHeading = (template, editPath) => {
  if (template.id && template.canEdit) {
    return (
      <a
        href={editPath.replace(':id', template.id)}
        target="_blank"
        rel="noopener noreferrer"
      >
        {template.name}
      </a>
    );
  }
  return template.name || ' ';
};

export const templateErrors = template => {
  if (Object.keys(aggregatedErrors(template)).length !== 0) {
    const res = Object.keys(aggregatedErrors(template)).map(key => (
      <li key={itemIteratorId(template, key)}>
        {formatError(key, aggregatedErrors(template)[key])}
      </li>
    ));
    return <ul>{res}</ul>;
  }
  return <span>There were no errors.</span>;
};

const aggregatedErrors = template => {
  const err = { ...template.errors } || {};
  if (template.additional_errors) {
    err.additional = template.additional_errors;
  }

  return err;
};

const formatError = (key, value) => {
  const omitKeys = ['base', 'additional'];
  if (omitKeys.reduce((memo, item) => memo || key === item, false)) {
    return value;
  }

  return `${key}: ${value}`;
};
