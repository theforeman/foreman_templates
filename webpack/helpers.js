import React from 'react';

import {
  propsToCamelCase,
} from 'foremanReact/common/helpers';

export const deepPropsToCamelCase = obj => {
  if (typeof obj !== 'object' || obj === null) {
    return obj;
  }
  if (Array.isArray(obj)) {
    return obj.map(deepPropsToCamelCase);
  }
  const transformed = propsToCamelCase(obj);
  return Object.keys(transformed).reduce((memo, key) => {
    memo[key] = deepPropsToCamelCase(transformed[key]);
    return memo;
  }, {})
};

// TODO: extract to core
export const withProtectedView = (
  ProtectedComponent,
  ProtectionComponent,
  protectionFn,
  extraProtectionProps = {}
) => props =>
  protectionFn(props) ? (
    <ProtectedComponent {...props} />
  ) : (
    <ProtectionComponent {...props} {...extraProtectionProps}/>
  );

export default withProtectedView;
