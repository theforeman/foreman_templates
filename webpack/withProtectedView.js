import React from 'react';

// TODO: extract to core
export const withProtectedView = (
  ProtectedComponent,
  ProtectionComponent,
  protectionFn,
  extraProtectionProps = {}
) => props =>
  protectionFn(props) ? (
    <ProtectedComponent {...props} />
  ) : (
    <ProtectionComponent {...props} {...extraProtectionProps} />
  );

export default withProtectedView;
