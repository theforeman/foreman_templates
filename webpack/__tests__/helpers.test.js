import React from 'react';
import { testSelectorsSnapshotWithFixtures } from 'react-redux-test-utils';
import { deepPropsToCamelCase, withProtectedView } from '../helpers';
import withProtectedView from '../withProtectedView';

const ProtectedComponent = () => <div>Protected component</div>;

const ProtectionComponent = () => <div>Protection component</div>;

const fixtures = {
  'should return protected component': () =>
    withProtectedView(ProtectedComponent, ProtectionComponent, () => true)(),
  'should return protection component': () =>
    withProtectedView(ProtectedComponent, ProtectionComponent, () => false)(),
};

describe('withProtectedView', () =>
  testSelectorsSnapshotWithFixtures(fixtures));

describe('deepPropsToCamelCase', () => {
  const deepProps = {
    my_key: [],
    array_key: ['a', 'b'],
    object_key: { nested_key: 'value' },
    null_key: null,
    undefined_key: undefined,
    zero_key: 0,
    string_key: '',
    string_key_again: 'test',
    nested_array: [{ random_key: 'random_value' }],
    not_number_key: NaN,
  };

  const deepCamelProps = {
    myKey: [],
    arrayKey: ['a', 'b'],
    objectKey: { nestedKey: 'value' },
    nullKey: null,
    undefinedKey: undefined,
    zeroKey: 0,
    stringKey: '',
    stringKeyAgain: 'test',
    nestedArray: [{ randomKey: 'random_value' }],
    notNumberKey: NaN,
  };

  it('should transform props', () => {
    expect(deepPropsToCamelCase(deepProps)).toEqual(deepCamelProps);
  });
});
