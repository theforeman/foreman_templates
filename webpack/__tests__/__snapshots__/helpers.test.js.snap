// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withProtectedView should return protected component 1`] = `<ProtectedComponent />`;

exports[`withProtectedView should return protection component 1`] = `<ProtectionComponent />`;
